dex/ws/wslink.go

```diff
@@ -100,15 +101,19 @@ func (c *WSLink) Send(msg *msgjson.Message) error {
 		return err
 	}
 
-	// If outChan is blocking (outBufferSize pending messages), many messages
-	// are being sent and not quickly enough. If the connection is actually
-	// down, c.quit will be closed. TODO: Add a default case with error return
-	// or just block until outHandler receives?
-	select {
-	case c.outChan <- b:
-	case <-c.quit:
-		return ErrPeerDisconnected
-	}
+	c.sendWG.Add(1)
+	c.outChan <- b
+	go func() {
+		c.writeMtx.Lock()
+		b := <-c.outChan
+		c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+		err := c.conn.WriteMessage(websocket.TextMessage, b)
+		c.writeMtx.Unlock()
+		c.sendWG.Done()
+		if err != nil {
+			c.Disconnect()
+		}
+	}()
```

Comment:
Although goroutines are not terribly heavy, we will be churning goroutines quite a bit this way.

To summarize what we're trying to accomplish here:  With a sending goroutine that receives on `outChan` as before we run the risk of filling that channel's buffer in `Send`.  Relying on a channel buffer was convenient, but Send might block if that buffer fills.  That maybe is not too bad because the `quit` change was the arbiter of connections status and messages were not dropped when the channel was full, just blocking.

We realized that a queue for the outgoing messages might needed.  The present solution solves the queuing issue via the Go scheduler, with a goroutine created for each message sent, at the cost of goroutine churn.  This may become a problem; I don't know.

Other solutions:
- A more complex implementation is to append under lock the messages to a slice of queued outgoing messages, while a `outHandler` goroutine pops entries off and sends them.  It's complex because a slice is not a synchronization mechanism.   The epochPump is an example, and that's pretty hairy.
- Have each `Send` block until it can actually `WriteMessage` itselft.  This would involve acquiring a semaphore (using a channel with buffer length 1, not a mutex, to select with quit in another case), and it would make `Send` blocking...

I actually think having `Send` block might be desirable so that the caller gets the error from `WriteMessage`, whereas now `nil` does not mean the send was successful.

Reply:
Anyway, we've chatted on matrix and I think we can run with this (a goroutine for each message sent) for now.  A requirement with `Send` is that a misbehaving client cannot gum up the works for up to `writeWait` (seconds).  Would we each connection to have a single goroutine for the life of the connection? Yes, but I think that using goroutines to handle this concurrency requirement is appropriate.